From fd5256d7cfae33cdf9a9f8b86f6c01480c7288b3 Mon Sep 17 00:00:00 2001
From: Zidan Wang <zidan.wang@freescale.com>
Date: Fri, 31 Jul 2015 17:02:48 +0800
Subject: [PATCH 0334/5242] MLK-11305 radio-si476x: support set
 V4L2_CID_AUDIO_MUTE CTRL

commit  455d3fb447ed6b63422428552eb9f9ba85e1d017 from
https://source.codeaurora.org/external/imx/linux-imx.git

By using gstreamer plugin v4l2radio, it will call VIDIOC_S_CTRL with
V4L2_CID_AUDIO_MUTE, but return failed.

So add V4L2_CID_AUDIO_MUTE CTRL support for radio-si476x.

Signed-off-by: Zidan Wang <zidan.wang@freescale.com>
(cherry picked from commit c18520adfd6de40dcc0659ddd778b0a2bd383cd4)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/radio/radio-si476x.c |   18 ++++++++++++++++++
 include/linux/mfd/si476x-core.h    |    2 ++
 2 files changed, 20 insertions(+)

diff --git a/drivers/media/radio/radio-si476x.c b/drivers/media/radio/radio-si476x.c
index d3609de..9880444 100644
--- a/drivers/media/radio/radio-si476x.c
+++ b/drivers/media/radio/radio-si476x.c
@@ -1012,6 +1012,14 @@ static int si476x_radio_s_ctrl(struct v4l2_ctrl *ctrl)
 		}
 		break;
 
+	case V4L2_CID_AUDIO_MUTE:
+		if (ctrl->val)
+			retval = regmap_write(radio->core->regmap,
+					      SI476X_PROP_AUDIO_MUTE, 3);
+		else
+			retval = regmap_write(radio->core->regmap,
+					      SI476X_PROP_AUDIO_MUTE, 0);
+		break;
 	default:
 		retval = -EINVAL;
 		break;
@@ -1531,6 +1539,16 @@ static int si476x_radio_probe(struct platform_device *pdev)
 		goto exit;
 	}
 
+	ctrl = v4l2_ctrl_new_std(&radio->ctrl_handler, &si476x_ctrl_ops,
+				 V4L2_CID_AUDIO_MUTE,
+				 0, 1, 1, 0);
+	rval = radio->ctrl_handler.error;
+	if (ctrl == NULL && rval) {
+		dev_err(&pdev->dev, "Could not initialize V4L2_CID_AUDIO_MUTE control %d\n",
+			rval);
+		goto exit;
+	}
+
 	if (si476x_core_has_diversity(radio->core)) {
 		si476x_ctrls[SI476X_IDX_DIVERSITY_MODE].def =
 			si476x_phase_diversity_mode_to_idx(radio->core->diversity_mode);
diff --git a/include/linux/mfd/si476x-core.h b/include/linux/mfd/si476x-core.h
index 674b45d..78a2b2c 100644
--- a/include/linux/mfd/si476x-core.h
+++ b/include/linux/mfd/si476x-core.h
@@ -493,6 +493,8 @@ enum si476x_common_receiver_properties {
 	SI476X_PROP_DIGITAL_IO_OUTPUT_SAMPLE_RATE	= 0x0202,
 	SI476X_PROP_DIGITAL_IO_OUTPUT_FORMAT		= 0x0203,
 
+	SI476X_PROP_AUDIO_MUTE				= 0x0301,
+
 	SI476X_PROP_SEEK_BAND_BOTTOM			= 0x1100,
 	SI476X_PROP_SEEK_BAND_TOP			= 0x1101,
 	SI476X_PROP_SEEK_FREQUENCY_SPACING		= 0x1102,
-- 
1.7.9.5

